EIPS/eip-7557.md

```diff
@@ -0,0 +1,370 @@
+Especially affected by this upcoming change will be transactions that involve smart contracts with a high code size.\
+Each such transaction in the block will be forced to pay the full "retail" price for loading smart contract
+bytecode during a transaction.
+
+The validators, however, only have to perform the actual reading from the Ethereum state once per block,
+and all subsequent reads of the values that were already referenced are significantly more efficient.\
+If witnesses are introduced to Ethereum blocks, the same witness can be reused by multiple transactions.\
+Forcing each transaction to pay regardless of the contents of the block is unfair and inefficient.
+
+Another change that may be on the roadmap for Ethereum is [RIP-7560: Native Account Abstraction](./rip-7560).\
+This change will see a large share of transactions being initiated by smart contracts directly.\
+It is reasonable to expect many of these Smart Contract Accounts to rely on the same core wallet implementations.\
+If each Account Abstraction transaction is charged a full gas cost of loading the Smart Contract Account code repeatedly,\
+such transactions would become significantly overpriced.\
+This difference is especially noticeable when compared to an EOA, which gets its validation logic loaded and executed for free.\
+In this scenario, the base gas fees would be taken from the senders and burned needlessly while block proposers
+would be enjoying an unjustified excessive earning of priority gas fees.
+
+## Specification
+
+The [EIP-2930: Optional access lists](./eip-2930) already introduced the first part of the solution.
+Each transaction can specify an array of `accessed_addresses` and `accessed_storage_keys` to announce its intention to
+read those values during the execution of the transaction.\
+The sender of the transaction is then pre-charged with the cost of accessing this data but is given a small discount
+compared to unannounced access.
+
+The missing component is a mechanism to aggregate the gas costs of the cold access and redistribute the resulting
+savings amongst the participating transactions.
+
+### Participant transactions mapping
+
+After the block builder finalizes the contents of the block, it iterates over all included transactions to read
+the `accessList` component of each supported transaction.
+
+The block builder then constructs an array containing each accessed address and each accessed slot, and an array of
+transaction senders' addresses that initiated at least one access to the given address or slot,
+as well as the `priorityFeePerGas` that was paid for such access.
+
+A sample JSON representation of the data structure that represents such a structure and is used in the pseudocode below:
+
+```json
+[
+  {
+    "address": "0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae",
+    "accessors": [
+      {
+        "sender": "0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1",
+        "priorityFeePerGas": "1000"
```

Comment:
although priorityFeePerGas is important to derive the reimbursements but they are again a picked from the txs, also sender info is not really important at this level. the compressed per tx reference can again resolve this